rust-runtime/aws-smithy-client/src/timeout.rs

```diff
@@ -51,8 +62,163 @@ impl Settings {
+    /// Given a function that will sleep a thread and timeout duration, create a new HttpRequestTimeout
+    /// that will timeout if the inner service doesn't respond before the timeout elapses.
+    pub fn new(inner: InnerService, duration: Option<Duration>) -> Self {
+        Self { inner, duration }
+    }
+
+    /// Create a new HttpRequestTimeout that will never timeout
+    pub fn no_timeout(inner: InnerService) -> Self {
+        Self {
+            inner,
+            duration: None,
+        }
+    }
+}
+
+/// A layer that wraps services in a timeout service
+#[non_exhaustive]
+#[derive(Debug)]
+pub struct TimeoutLayer(Option<Duration>);
+
+impl TimeoutLayer {
+    /// Create a new HttpRequestTimeoutLayer
+    pub fn new(duration: Option<Duration>) -> Self {
+        TimeoutLayer(duration)
+    }
+}
+
+impl<InnerService> Layer<InnerService> for TimeoutLayer {
+    type Service = TimeoutService<InnerService>;
+
+    fn layer(&self, inner: InnerService) -> Self::Service {
+        TimeoutService {
+            inner,
+            duration: self.0,
+        }
+    }
+}
+
+pin_project! {
+    #[non_exhaustive]
+    #[must_use = "futures do nothing unless you `.await` or poll them"]
+    /// A future representing a timeout timer. Wraps a [Timeout] with extra context for error reporting
+    pub struct TimeoutLayerFuture<T> {
+        #[pin]
+        inner: Timeout<T, Sleep>
+    }
+}
+
+impl<InnerFuture, T, E> Future for TimeoutLayerFuture<InnerFuture>
+where
+    InnerFuture: Future<Output = Result<T, SdkError<E>>>,
+{
+    type Output = Result<T, SdkError<E>>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let me = self.project();
+        match me.inner.poll(cx) {
+            Poll::Ready(Ok(t)) => Poll::Ready(t),
+            Poll::Ready(Err(timeout)) => {
+                Poll::Ready(Err(SdkError::ConstructionFailure(timeout.into())))
+            } // TODO: need to put this in the right place
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+impl<H, R, InnerService, E> tower::Service<Operation<H, R>> for TimeoutService<InnerService>
+where
+    InnerService: tower::Service<Operation<H, R>, Error = SdkError<E>>,
+{
+    type Response = InnerService::Response;
+    type Error = aws_smithy_http::result::SdkError<E>;
+    type Future = TimeoutLayerFuture<InnerService::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Operation<H, R>) -> Self::Future {
+        let base_future = self.inner.call(req);
+        let timeout_future = self.duration.map_or_else(
+            // If no timeout duration is provided, create a functionally infinite one
+            || default_async_sleep().expect("No default sleep impl was found. This is unexpected. Please report this bug.").sleep(Duration::MAX),
+            |duration| TokioSleep::new().sleep(duration),
```

Comment:
this isn't quite right since it will panic of `tokio-rt` isn't enabled. Probably need to either:
- make the `Timeout` future optional
- use `MaybeTimeoutFuture`